rfc/system/5741-binary-idl-with-message-pack.md

```diff
@@ -0,0 +1,525 @@
+     `dict` -> `MessagePack Bytes` -> `Binary (value: MessagePack Bytes, tag: msgpack) IDL Object`.
+
+2. There is **NO JSON** involved in the new type at all. Only **JSON Schema** is used to construct `DataClass` or `Pydantic BaseModel`.
+
+
+### Literal Type
+Literal Type will be `Protobuf struct`.
+`Json Schema` will be stored in `Literal Type's metadata`.
+
+1. Dataclass, Pydantic BaseModel and pure dict in python will all use `Protobuf Struct`.
+2. We will put `Json Schema` in Literal Type's `metadata` field, this will be used in flytekit remote api to construct dataclass/Pydantic BaseModel by `Json Schema`.
+3. We will use libraries written in golang to compare `Json Schema` to solve this issue: ["[BUG] Union types fail for e.g. two different dataclasses"](https://github.com/flyteorg/flyte/issues/5489).
+
+
+Note: The `metadata` of `Literal Type` and `Literal Value` are not the same.
+
+## 2 Motivation
+
+In Flytekit, when handling dataclasses, Pydantic base models, and dictionaries, we store data using a JSON string within Protobuf struct datatype.
+
+This approach causes issues with integers, as Protobuf struct does not support int types, leading to their conversion to floats.
+
+This results in performance issues since we need to recursively iterate through all attributes/keys in dataclasses and dictionaries to ensure floats types are converted to int.
+
+In addition to performance issues, the required code is complicated and error prone.
+
+Note: We have more than 10 issues about dict, dataclass and Pydantic.
+
+This feature can solve them all.
+
+## 3 Proposed Implementation
+### Before
+```python
+@task
+def t1() -> dict:
+  ...
+  return {"a": 1} # Protobuf Struct {"a": 1.0}
+
+@task
+def t2(a: dict):
+  print(a["integer"]) # wrong, will be a float
+```
+### After
+```python
+@task
+def t1() -> dict: # Literal(scalar=Scalar(binary=Binary(value=b'msgpack_bytes', tag="msgpack")))
+  ...
+  return {"a": 1}  # Protobuf Binary value=b'\x81\xa1a\x01', produced by msgpack
+
+@task
+def t2(a: dict):
+  print(a["integer"]) # correct, it will be a integer
+```
+
+#### Note
+- We will use implement `to_python_value` to every type transformer to ensure backward compatibility.
+For example, `Binary IDL Object` -> python value and `Protobuf Struct IDL Object` -> python value are both supported.
+
+### How to turn a value to bytes?
+#### Use MsgPack to convert a value into bytes
+##### Python
+```python
+import msgpack
+
+# Encode
+def to_literal():
+  msgpack_bytes = msgpack.dumps(python_val)
+  return Literal(scalar=Scalar(binary=Binary(value=b'msgpack_bytes', tag="msgpack")))
+
+# Decode
+def to_python_value():
+    # lv: literal value
+    if lv.scalar.binary.tag == "msgpack":
+        msgpack_bytes = lv.scalar.json.value
+    else:
+        raise ValueError(f"{tag} is not supported to decode this Binary Literal: {lv.scalar.binary}.")
+    return msgpack.loads(msgpack_bytes)
```

Comment:
1. We don't need this, because we will use `from mashumaro.codecs.msgpack import MessagePackDecoder, MessagePackEncoder` to encode and decode.
It will make sure we convert it back to a type we 100% want.

2. `msgpack.dumps` will only be used when dealing with `untyped dict`.